python/hyperon/atoms.py

```diff
@@ -0,0 +1,180 @@
+    VARIABLE = Atom._from_catom(hp.CAtomType.VARIABLE)
+    EXPRESSION = Atom._from_catom(hp.CAtomType.EXPRESSION)
+    GROUNDED = Atom._from_catom(hp.CAtomType.GROUNDED)
+
+class GroundedAtom(Atom):
+
+    def __init__(self, catom):
+        super().__init__(catom)
+
+    def get_object(self):
+        return hp.atom_get_object(self.catom)
+
+    def get_grounded_type(self):
+        return Atom._from_catom(hp.atom_get_grounded_type(self.catom))
+
+def G(object, type=AtomType.UNDEFINED):
+    return GroundedAtom(hp.atom_gnd(object, type.catom))
+
+def call_execute_on_grounded_atom(gnd, typ, args):
+    # ... if hp.atom_to_str(typ) == AtomType.UNDEFINED
+    res_typ = AtomType.UNDEFINED if hp.atom_get_type(typ) != AtomKind.EXPR \
+        else Atom._from_catom(hp.atom_get_children(typ)[-1])
+    args = [Atom._from_catom(catom) for catom in args]
+    return gnd.execute(*args, res_typ=res_typ)
+
+def atoms_are_equivalent(first, second):
+    return hp.atoms_are_equivalent(first.catom, second.catom)
+
+class GroundedObject:
+
+    def __init__(self, content, id=None):
+        self.content = content
+        self.id = id
+
+    def __repr__(self):
+        return str(self.content) if self.id is None else self.id
+
+    def copy(self):
+        return self
+
+class ValueObject(GroundedObject):
+
+    @property
+    def value(self):
+        return self.content
+
+    def __eq__(self, other):
+        # TODO: ?typecheck
+        return isinstance(other, ValueObject) and self.content == other.content
+
+class NoReduceError(Exception):
+    pass
+
+class OperationObject(GroundedObject):
+
+    def __init__(self, name, op, unwrap=True):
+        super().__init__(op, name)
+        self.unwrap = unwrap
+
+    @property
+    def op(self):
+        return self.content
+
+    @property
+    def name(self):
+        return self.id
+
+    def execute(self, *args, res_typ=AtomType.UNDEFINED):
+        # type-check?
+        if self.unwrap:
+            for arg in args:
+                if not isinstance(arg, GroundedAtom):
+                    # REM:
+                    # Currently, applying grounded operations to pure atoms is not reduced.
+                    # If we want, we can raise an exception, or to form a error expression instead,
+                    # so a MeTTa program can catch and analyze it.
+                    # raise RuntimeError("Grounded operation " + self.name + " with unwrap=True expects only grounded arguments")
+                    raise NoReduceError()
+            args = [arg.get_object().content for arg in args]
+            return [G(ValueObject(self.op(*args)), res_typ)]
+        else:
+            result = self.op(*args)
+            if not isinstance(result, list):
+                raise RuntimeError("Grounded operation `" + self.name + "` should return list")
+            return result
+
+    def __eq__(self, other):
+        return isinstance(other, OperationObject) and self.name == other.name
+
+def type_sugar(type_names):
```

Comment:
Minor and I know it is not directly related to the PR description, but I would prefix `type_sugar` and `call_execute_on_grounded_atom` by `_` to remove them from public API of the module.

Reply:
`call_execute_on_grounded_atom` is used in `hyperonpy.cpp`, so it should be in the public API

Reply:
I've change `type_sugar`. It might be used outside, though...

Reply:
Nothing should prevent usage of `_call_execute_on_grounded_atom` from `hyperonpy.cpp`, it just will say to users that they should not use it directly.